installinator/src/dispatch.rs

```diff
@@ -0,0 +1,150 @@
+
+use anyhow::{Context, Result};
+use buf_list::BufList;
+use camino::{Utf8Path, Utf8PathBuf};
+use clap::{Args, Parser, Subcommand};
+use slog::Drain;
+use tokio::io::AsyncWriteExt;
+
+use crate::{
+    errors::DiscoverPeersError,
+    peers::{ArtifactId, FetchedArtifact, Peers},
+};
+
+/// Installinator app.
+#[derive(Debug, Parser)]
+#[command(version)]
+pub struct InstallinatorApp {
+    #[clap(subcommand)]
+    subcommand: InstallinatorCommand,
+}
+
+impl InstallinatorApp {
+    /// Executes the app.
+    pub async fn exec(self) -> Result<()> {
+        let log = Self::setup_log("/tmp/installinator.log").unwrap();
+
+        match self.subcommand {
+            InstallinatorCommand::Discover(opts) => opts.exec(log).await,
+            InstallinatorCommand::Install(opts) => opts.exec(log).await,
+        }
+    }
+
+    fn setup_log(path: impl AsRef<Utf8Path>) -> anyhow::Result<slog::Logger> {
+        let file = std::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(path.as_ref())?;
+
+        let file_decorator = slog_term::PlainDecorator::new(file);
+        let file_drain =
+            slog_term::FullFormat::new(file_decorator).build().fuse();
+
+        let stderr_decorator =
+            slog_term::PlainDecorator::new(std::io::stderr());
+        let stderr_drain =
+            slog_term::FullFormat::new(stderr_decorator).build().fuse();
+
+        let drain = slog::Duplicate::new(file_drain, stderr_drain).fuse();
+        let drain = slog_async::Async::new(drain).build().fuse();
+        Ok(slog::Logger::root(drain, slog::o!()))
+    }
+}
+
+#[derive(Debug, Subcommand)]
+enum InstallinatorCommand {
+    /// Discover peers on the bootstrap network.
+    Discover(DiscoverOpts),
+    /// Perform the installation.
+    Install(InstallOpts),
+}
+
+#[derive(Debug, Args)]
+#[command(version)]
+struct DiscoverOpts {
+    //
+}
+
+impl DiscoverOpts {
+    async fn exec(self, log: slog::Logger) -> Result<()> {
+        let peers = Peers::mock_discover(&log).await?;
+        println!("discovered peers: {}", peers.display());
+        Ok(())
+    }
+}
+
+#[derive(Debug, Args)]
+#[command(version)]
+struct InstallOpts {
+    // TODO: fetch this
+    artifact_id: ArtifactId,
+
+    // TODO: checksum?
+
+    // The destination to write to.
+    destination: Utf8PathBuf,
+}
+
+impl InstallOpts {
+    async fn exec(self, log: slog::Logger) -> Result<()> {
+        // TODO: is buffering up the entire artifact in memory OK?
+        let artifact = FetchedArtifact::loop_fetch_from_peers(
+            &log,
+            || {
+                // TODO: discover nodes via the bootstrap network
+                async {
+                    Peers::mock_discover(&log)
+                        .await
+                        .context("error discovering peers")
+                        .map_err(DiscoverPeersError::Retry)
+                }
```

Comment:
1. I've been thinking of installinator as doing everything sled-agent (or really bootstrap agent) would do to get to the point of communicating on the bootstrap network, so I'm inclined to say yes, the installinator should launch maghemite (since normally bootstrap-agent [launches it](https://github.com/oxidecomputer/omicron/blob/ac88471f4c20540122b5ced344cede295b8c41bd/sled-agent/src/bootstrap/maghemite.rs#L29)). I'm fuzzy on other requirements to get to the bootstrap network - at a minimum, starting maghemite requires passing it the names of the interfaces it should be using, so those have to exist (and possibly have link-local addresses already?).

2. I was testing locally on a Linux host with multiple [helios-engvm](https://github.com/oxidecomputer/helios-engvm) VMs talking to each other. This required a bit of finagling (although not as much as I expected, I think) with libvirt networking settings to make sure the VMs could see each other. It's been many months since I tried that, but it might still work. After getting the VMs starting up and setting up link-local addresses, I had to add routes out in the linux host to let them see each other. For the two VMs I had running and the virtual bridge device libvirt created, that was:

```
ip -6 route add fdb0:5254:e4:589f::1/64 via 1de::1f7 dev virbr1
ip -6 route add fdb0:5254:8a:8c1c::1/64 via 1de::1e6 dev virbr1
```

This felt fragile at the time and still does. I'm not sure if there's a better way to fake a bootstrap network locally for testing though.

Reply:
You should be able to use [falcon](https://github.com/oxidecomputer/falcon) for this if you have a helios box. I created an [example two-node toplolgy](https://github.com/oxidecomputer/falcon-omicron) that I used to test trust quourm. Unfortunately (like the falcon instructions)t his is somewhat out of date. Primarily I think this is just because of the images that should be used. Ry or Levon should be able to help you here. I think Falcon is a much more reliable way to do this type of testing personally.

Reply:
Thanks everyone. I do have a Helios VM -- I'm going to play around with falcon.

Reply:
Managed to get a setup going on atrium with a ton of help from @rcgoodfellow. I've updated the PR with basic discovery that seems to work, will make some notes on the updated version.

Reply:
So I basically got this working by:

**Using physical hardware rather than a VM**, because there appear to be issues with kvm/nested virtualization on Linux. I used atrium for this.

**Using a falcon template** which copies over a ton of scripts from falcon-omicron: https://github.com/sunshowers/falcon-bootstrap

I didn't use falcon-omicron because I couldn't get it to build, so I started from a new template and copied over the config and scripts from falcon-omicron.

**Changing the firstboot.sh script to statically assign IPs**, since DHCP wasn't working for me in the falcon environment on atrium (could be just an IP misconfiguration somewhere):

```bash
case $(hostname) in
        rain_om1)
                v4_addr="192.168.3.101"
                ;;
        rain_om2)
                v4_addr="192.168.3.102"
                ;;
        *)
                echo -n "unknown hostname $(hostname)"
                exit 1
                ;;
esac
```

**Downloading ddmadm and ddmd** from [this CI job](https://github.com/oxidecomputer/maghemite/runs/9492743660) and scping them over.

**Spinning up ddmd** in both VMs with `pfexec ./ddmd 8000 :: vioif0/v6 server`.

**Using ddmadm get-peers** to verify that both ddmd instances could see each other:

```
rain@rain_om2:~$ ./ddmadm get-peers
{
    "3": PeerInfo {
        addr: fe80::8:20ff:fe9e:326e,
        host: Some(
            "rain_om1",
        ),
        kind: Some(
            Server,
        ),
        status: Active,
    },
}
```

**Creating addresses and advertising them**:

* on rain-om1: `ipadm create-addr -T static -a fd00:1::1/64 vioif0/u6 && ddmadm advertise-prefix fd00:1::1/64`
* on rain-om2: `ipadm create-addr -T static -a fd00:2::1/64 vioif0/u6 && ddmadm advertise-prefix fd00:2::1/64`

(There was an issue here with the ddmds not adding these addresses to the kernel routing tables as inspected with `netstat -nr -f inet6`, which was fixed by restarting the ddmd servers.)

**Copying over wicketd to rain-om1 and running it** with `./wicketd run --address '[fd00:1::1]:14000' ~/config.toml`

**Copying over installinator to rain-om2 and running it** with `./installinator install __installinator-test:100000000 test.out`

This copied over a 100MB file from wicketd as expected. (Yay!)